Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp

```diff
@@ -0,0 +1,546 @@
+            return Type::UTF8;
+        default:
+            return Error::from_string_literal("TIFFImageDecoderPlugin: Unknown type");
+        }
+    }
+
+    static constexpr u8 size_of_type(Type type)
+    {
+        switch (type) {
+        case Type::Byte:
+            return 1;
+        case Type::ASCII:
+            return 1;
+        case Type::Short:
+            return 2;
+        case Type::Long:
+            return 4;
+        case Type::Rational:
+            return 8;
+        case Type::Undefined:
+            return 1;
+        case Type::SLong:
+            return 4;
+        case Type::SRational:
+            return 8;
+        case Type::UTF8:
+            return 1;
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    ErrorOr<Vector<Value, 1>> read_tiff_value(Type type, u32 count, u32 offset)
+    {
+        auto const old_offset = TRY(m_stream->tell());
+        ScopeGuard reset_offset { [this, old_offset]() { MUST(m_stream->seek(old_offset)); } };
+
+        TRY(m_stream->seek(offset));
+
+        auto const read_each_value = [this, count]<typename T>() -> ErrorOr<Vector<Value>> {
+            Vector<Value, 1> result {};
+            TRY(result.try_ensure_capacity(count));
+            if constexpr (IsSpecializationOf<T, Rational>) {
+                for (u32 i = 0; i < count; ++i)
+                    result.empend(T { TRY(read_value<typename T::type>()), TRY(read_value<typename T::type>()) });
+            } else {
+                for (u32 i = 0; i < count; ++i)
+                    result.empend(TRY(read_value<T>()));
+            }
+            return result;
+        };
+
+        switch (type) {
+        case Type::Byte:
+        case Type::Undefined:
+            return read_each_value.template operator()<u8>();
+        case Type::ASCII:
+        case Type::UTF8: {
+            Vector<Value, 1> result;
+            auto string_data = TRY(ByteBuffer::create_uninitialized(count));
+            TRY(m_stream->read_until_filled(string_data));
+            result.empend(TRY(String::from_utf8(StringView { string_data.bytes() })));
+            return result;
+        }
+        case Type::Short:
+            return read_each_value.template operator()<u16>();
+        case Type::Long:
+            return read_each_value.template operator()<u32>();
+        case Type::Rational:
+            return read_each_value.template operator()<Rational<u32>>();
+        case Type::SLong:
+            return read_each_value.template operator()<i32>();
+            ;
+        case Type::SRational:
+            return read_each_value.template operator()<Rational<i32>>();
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    }
+
+    ErrorOr<void> read_tag()
+    {
+        auto const tag = TRY(read_value<u16>());
+        auto const type = TRY(read_type());
+        auto const count = TRY(read_value<u32>());
+
+        auto const total_data_size = size_of_type(type) * count;
```

Comment:
Can we return an error here if `total_data_size >= m_stream->remaining()`?

Also `total_data_size` can currently overflow a u32.

Reply:
Do you still have the test case? I don't understand why do we need this check, using a `Stream` should allow us to discard it. If we don't have enough data in the file, one of the call to `read` or `seek` should return an error.

Reply:
Sure, [here it is](https://github.com/SerenityOS/serenity/files/13198637/tiff_fuzzer_oom.zip). 

Sorry, I just realized I didn't actually explain what the problem was. The test case has a `total_data_size` larger than 2GB (the fuzzer's default OOM limit).  The OOM occurs on line 246.

Reply:
Ok, make sense now. And we should indeed catch that!